### PR TITLE
Admin panel

### PR DESCRIPTION
**Pull request by Kaszana102**

### CODE REVIEW

File: test/user_dto_test.py
Comment (MikolekN) [inline test/user_dto_test.py:1]: Unnecessary change.

File: src/user_dto.py
Comment (MikolekN) [inline src/user_dto.py:1]: Unnecessary change.

File: functional/functional_test_constant.py
Comment (MikolekN) [inline functional/functional_test_constant.py:1]: What is the need for this file?

Comment (Kaszana102) [reply functional/functional_test_constant.py:1]: I was told to the test for the admin panel. So they only check if button works, to check the notificiation of clicked button. There is no way to test sending mail, and invoice this way.

File: src/user_dto.py
Comment (Kaszana102) [reply src/user_dto.py:1]: actually some of it was needed. Firstly at the beginning i hadn't fully understood how to do we treat bundles and subscription, so I changed the text accordingly. Payment method is nowhere stored, so i removed if form the text.

File: test/user_dto_test.py
Comment (Kaszana102) [reply test/user_dto_test.py:1]: i generating text was modified, i had also to modify the tests :/

File: functional/functional_test_constant.py
Comment (MikolekN) [reply functional/functional_test_constant.py:1]: What I meant is that from what I can see the file is only used in admin_panel_test.py where it's used to set a value for self.BASE_URL. Why not set it the same way as in other tests?
Instead of: self.BASE_URL = BASE_URL + '/admin_panel'
Use: self.BASE_URL = os.getenv("BASE_URL")
And then:
Instead of: self.driver.get(self.BASE_URL)
Use: self.driver.get((os.path.join(self.BASE_URL, "admin_panel")))

Comment (Kaszana102) [reply functional/functional_test_constant.py:1]: @BognaLew told me to create some file with BASE_URL, instead of geting url form env.

Comment (MikolekN) [reply functional/functional_test_constant.py:1]: ok

File: test/user_dto_test.py
Comment (MikolekN) [reply test/user_dto_test.py:1]: These tests are unnecessary because they duplicate the code of other tests:
test_xml_serialization_deserialization_empty_packets is the same as test_xml_serialization_deserialization
test_csv_serialization_deserialization_empty_packets is the same as test_csv_serialization_deserialization

File: src/user_dto.py
Comment (MikolekN) [reply src/user_dto.py:1]: The additional text in the comment isn't necessary because it doesn't accurately reflect the code.

File: test/user_dto_test.py
Comment (Kaszana102) [reply test/user_dto_test.py:1]: oh wait, i looked at the wrong file. Ok right it's useless. I might not have  reverted oll useless changes. Tests removed

File: src/user_dto.py
Comment (Kaszana102) [reply src/user_dto.py:1]: jeez, i looked at the file above the comment, not below. ( i looked at generator_text.py)
The comment is also a part of history.  Removed.

**Review by MikolekN — APPROVED** [review]

File: src/csvDatabase.py [hunk 1]
```diff
@@ -71,10 +71,15 @@ def add_client(self, client_data: Dict) -> None:
...
             raise KeyError(f"Client with id {client_data['id']} already exists.")
 
+        client_data_copy = client_data.copy()  # prevent data change writing a list
         with open(self._filename, 'a', newline='') as csvfile:
             writer = csv.DictWriter(csvfile, fieldnames=self._fields)
 
-            writer.writerow(client_data)
+            for key in client_data_copy.keys():
+                if type(client_data_copy[key]) is list:
+                    client_data_copy[key] = '['+'.'.join(client_data_copy[key])+']'
```
Comment (Lenetis) [inline src/csvDatabase.py]: 1. What if `type(client_data_copy[key])` is a `tuple` and not a `list`?
2. Why convert lists to string in a manual way, instead of just `str(list)`?
3. Why is this list separated by dots and not commas?

File: src/csvDatabase.py [hunk 2]
```diff
@@ -86,6 +91,14 @@ def get_client(self, client_id: str) -> Dict | None:
             reader = csv.DictReader(csvfile, fieldnames=self._fields)
             for data in reader:
                 if data["id"] == client_id:
+                    #check for lists values
+                    for key in data:
+                        if '[' in data[key]:
+                            data[key] = data[key].replace('[', '').replace(']', '')
```
Comment (Lenetis) [inline src/csvDatabase.py]: Why convert string to list in a manual way?
If you converted lists to strings with just `string = str(list)`, you could use `list = ast.literal_eval(string)` here.

File: src/csvDatabase.py [hunk 3]
```diff
@@ -150,6 +163,14 @@ def get_clients(self, predicate: Callable[[Dict], bool] = None) -> List:
             next(reader)  # skip headers row
             for data in reader:
                 if predicate is None or predicate(data):
+                    # check for lists values
+                    for key in data:
+                        if '[' in data[key]:
+                            data[key] = data[key].replace('[', '').replace(']', '')
```
Comment (Lenetis) [inline src/csvDatabase.py]: Manual conversion of string to list, same comment as before.

File: src/database.py [hunk 4]
```diff
@@ -14,33 +20,63 @@ def get_user_sex(self, user_id) -> str:
...
         """
 
-        # TO IMPLEMENT
+        client = self.db.basic_db.get_client(user_id)
 
-        return ''
+        if client['gender'] == 'female':
+            return 'F'
+        else:
+            return 'M'
```
Comment (Lenetis) [inline src/database.py:27]: The registration form also has "Other" gender

File: src/database.py [hunk 5]
```diff
@@ -14,33 +20,63 @@ def get_user_sex(self, user_id) -> str:
...
-        # TO IMPLEMENT
+        email = self.db.basic_db.get_client(user_id)['email']
+        client = self.db.get_client_by_email(email)
+        subscribed = client.bundles
+        not_subscribed = []
+
+        subscribed_packets_names = [x.name for x in subscribed]
 
-        return []
+        for service in PACKETS.keys():
```
Comment (Lenetis) [inline src/database.py]: Little nitpick, but why "for **service** in PACKETS.keys()"? Woudn't "for **packet** in PACKETS.keys()" be more fitting?

File: src/database.py [hunk 6]
```diff
@@ -14,33 +20,63 @@ def get_user_sex(self, user_id) -> str:
...
+            if service not in subscribed_packets_names:
+                not_subscribed.append(PACKETS[service]['name'])
+
+        return not_subscribed
+
+    def get_subscription(self, user_id) -> str:
+        """
+        Function which retrieves all services subscribed by user
+        :param self:
```
Comment (Lenetis) [inline src/database.py]: `:param self` is unnecessary in a docstring.
`user_id` should have a typehint.

File: src/mail.py [hunk 7]
```diff
@@ -31,22 +31,22 @@ def send_mail(receiver: str, subject: str, content: str, files: List[str]) -> No
...
-    for filename in files or []:
+    for filename in files:
         with open(filename, "rb") as file:
             part = MIMEApplication(
                 file.read(),
                 Name=basename(filename)
             )
         # After the file is closed
-        part['Content-Disposition'] = 'attachment; filename="{basename(f)}"'
+        part['Content-Disposition'] = 'attachment; filename="%s"' % basename(filename)
```
Comment (Lenetis) [inline src/mail.py:49]: FYI, the previous f-string didn't work because it lacked an `f` in front.
`f'attachment; filename="{basename(f)}"'` would work fine.

File: src/main.py [hunk 8]
```diff
@@ -1,9 +1,21 @@
 import datetime
+
+import inspect
+from dataclasses import asdict
+import os
```
Comment (Lenetis) [inline src/main.py:3]: - Wrong import order (imports within one grup should be in alphabetic order, and `import ...` should be before `from ... import`)
- Unnecessary newline after datetime import
- inspect and asdict import statements are unused. Also generate_invoice_xml and process_form which are in the next few lines

https://peps.python.org/pep-0008/#imports

File: src/main.py [hunk 9]
```diff
@@ -156,3 +172,37 @@ def edit_subscription():
...
+                          mail_text.__str__(),
+                          [])
+            return render_template("admin_panel.html", the_title="Paint Drying", notification="Mails sent")
+        if request.args.get('send-invoice') == 'send':
+            clients = db.basic_db.get_clients()
+            if len(clients) > 0:
+                client = db.get_client_by_email(clients[0]['email'])
+                file_name = 'invoice.pdf'
+                invoice = Invoice(client)
+                invoice.save_pdf('invoice.pdf')
```
Comment (Lenetis) [inline src/main.py]: In line 194 you define a variable called "`file_name`", but here the filename is given as a string

File: src/templates/admin_panel.html [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+<h1>DC Paint Drying</h1>
+<a href="..">Main Page</a>
+
+<p style="text-align:center; margin: 100px; font-size: 70px">Admin Panel</p>
+<form method="get" class="vertical-flex">
+	<div class="horizontal-flex">
+		<div class="admin-button">
+			<input type="submit" value="suggest" name="suggest-services" id="suggest-services">
+			<label for="suggest-services" class="vertical-flex" style="height: 100%;" id="suggest-services-label">
```
Comment (Lenetis) [inline src/templates/admin_panel.html:13]: Why use `style="..."` in html instead of using css?

File: src/templates/admin_panel.html [hunk 11]
```diff
@@ -0,0 +1,39 @@
+{% extends 'base.html' %}
+
+{% block content %}
+
+<h1>DC Paint Drying</h1>
+<a href="..">Main Page</a>
```
Comment (Lenetis) [inline src/templates/admin_panel.html:6]: In `index.html` you use `<a href="{{ url_for('admin_panel') }}">`. Could a similar method be used here, instead of the ".."?

File: src/text_generator.py [hunk 12]
```diff
@@ -1,7 +1,9 @@
+from src import manifest
```
Comment (Lenetis) [inline src/text_generator.py]: Unused import statement

File: src/user_dto.py [hunk 13]
```diff
@@ -70,4 +70,5 @@ def from_csv(cls, csv_str):
         for i in range(len(data)):
             if data[i].isdigit():
                 data[i] = int(data[i])
+
```
Comment (Lenetis) [inline src/user_dto.py:73]: Unnecessary newline

File: test/csv_test.py [hunk 14]
```diff
@@ -174,3 +174,45 @@ def test_use_existing_csv_file_with_auto_columns():
...
+    client = {"id": "01", "packets": ['a','b']}
+
+    db1 = CSVDatabase("dbtest.csv", columns)
+    db1.add_client(client)
+
+    db2 = CSVDatabase("dbtest.csv")
+    assert db2.get_client("01") == client
+
+    db1.drop_database()
+def test_get_all_clients_with_list():
```
Comment (Lenetis) [inline test/csv_test.py]: Rename to `test_get_all_clients_with_packets_list` (other tests you added follow this convention)

File: test/database_test.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
+import unittest
+import uuid
+
+from src.bundle_info import BundleInfo
+from src.client_info import ClientInfo
+from src.csvDatabase import CSVDatabase
+from src.database import Database
+from src.database_context import DatabaseContext
+from src.subscription_info import SubscriptionInfo
+from src.user_dto import UserDto
```
Comment (Lenetis) [inline test/database_test.py]: Most of these import statements are unused.

File: test/test_text_generator.py [hunk 16]
```diff
@@ -1,102 +1,118 @@
...
+        self.context.destroy()
     def test_greeting(self):
-        database = Database()
-        database.get_user_sex = MagicMock(return_value='M')
-        database.get_user_surname = MagicMock(return_value='Wierzba')
+        self.db.get_user_sex = MagicMock(return_value='M')
+        self.db.get_user_surname = MagicMock(return_value='Wierzba')
 
-        assert text_gen.replace_keywords('{$greeting}', 1, database) == 'Szanowny Panie Wierzba'
+        assert text_gen.replace_keywords('{$greeting}', 1, self.db) == 'Szanowny Panie Wierzba'
```
Comment (Lenetis) [inline test/test_text_generator.py]: `text_gen.replace_keywords('{$greeting}', 1, self.db)`
`Expected type 'str', got 'int' instead---^`

Same problem happens in other test

File: functional/functional_test_constant.py
Comment (BognaLew) [reply functional/functional_test_constant.py:1]: I think you misunderstood me. You should be using environmental variable, but you need to remember to set it in PyCharm in run configuration. Please switch to environmental variable before merge.

Comment (Kaszana102) [reply functional/functional_test_constant.py:1]: switched

File: src/csvDatabase.py (hunk 1 above)
Comment (Kaszana102) [reply src/csvDatabase.py]: 0. removed lists in csv (Bartek made managing packets in a more sensible way)
2. because commas
3. because commas (csv was comma separated)

File: src/csvDatabase.py (hunk 2 above)
Comment (Kaszana102) [reply src/csvDatabase.py]: removed funcionalisty

File: src/csvDatabase.py (hunk 3 above)
Comment (Kaszana102) [reply src/csvDatabase.py]: removed lists in csv functionality

File: src/database.py (hunk 4 above)
Comment (Kaszana102) [reply src/database.py:27]: how i am supposed to write mail to nonbinary person?

File: src/database.py (hunk 5 above)
Comment (Kaszana102) [reply src/database.py]: renames

File: src/text_generator.py (hunk 12 above)
Comment (Kaszana102) [reply src/text_generator.py]: removed line

File: src/user_dto.py (hunk 13 above)
Comment (Kaszana102) [reply src/user_dto.py:73]: literaly pycharm sees it as a problem, when there is no newline there XD

File: test/csv_test.py (hunk 14 above)
Comment (Kaszana102) [reply test/csv_test.py]: removed tests

File: test/test_text_generator.py (hunk 16 above)
Comment (Kaszana102) [reply test/test_text_generator.py]: changed

File: src/user_dto.py (hunk 13 above)
Comment (Lenetis) [reply src/user_dto.py:73]: Ok, in that case I retract my comment. I joust found it strange that the only change you made to this file was adding a newline, and I thought it was a mistake

**Review by Lenetis — APPROVED** [review]

**Review by Lenetis — APPROVED** [review]
